lib/src/main/kotlin/com/swmansion/starknet/data/types/Contracts.kt

```diff
@@ -109,24 +116,222 @@ data class ContractDefinition(private val contract: String) {
+                    .lineSequence().map { it.trim() }.joinToString("\n")
+                    .replace("\n", "")
+                    .replace(",\"", ", \"")
+                    .replace(",{", ", {")
+                    .replace(",[", ", ["),
+            ) else put("abi", "")
+        }
+    }
+}
+
+@Serializable
+data class CasmContractDefinition(private val contract: String) {
+    private val casmClassVersion: JsonElement = JsonPrimitive("COMPILED_CLASS_V1")
+    private val prime: JsonElement
+    private val hints: JsonElement
+    private val compilerVersion: JsonElement
+    private val entryPointsByType: JsonElement
+    private val bytecode: JsonElement
+
+    class InvalidContractException(missingKey: String) :
+        RuntimeException("Attempted to parse an invalid contract. Missing key: $missingKey")
+
+    init {
+        val (entryPointsByType, bytecode, prime, hints, compilerVersion) = parseContract(contract)
+        this.entryPointsByType = entryPointsByType
+        this.bytecode = bytecode
+        this.prime = prime
+        this.hints = hints
+        this.compilerVersion = compilerVersion
+    }
+
+    private fun parseContract(contract: String): Array<JsonElement> {
+        val compiledContract = Json.parseToJsonElement(contract).jsonObject
+        val entryPointsByType =
+            compiledContract["entry_points_by_type"] ?: throw InvalidContractException("entry_points_by_type")
+        val bytecode = compiledContract["bytecode"] ?: throw InvalidContractException("bytecode")
+        val prime = compiledContract["prime"] ?: throw InvalidContractException("prime")
+        val hints = compiledContract["hints"] ?: throw InvalidContractException("hints")
+        val compilerVersion = compiledContract["compiler_version"] ?: throw InvalidContractException("compiler_version")
+        return arrayOf(entryPointsByType, bytecode, prime, hints, compilerVersion)
+    }
+
+    fun toJson(): JsonObject {
+        return buildJsonObject {
+            put("casm_class_version", casmClassVersion)
+            put("entry_points_by_type", entryPointsByType)
+            put("bytecode", bytecode)
+            put("prime", prime)
+            put("hints", buildJsonArray { hints.toString() })
+            put("compiler_version", compilerVersion)
+        }
+    }
+}
+
+sealed class ContractClassBase
+
+@Serializable
+data class DeprecatedContractClass(
     val program: String,
 
     @SerialName("entry_points_by_type")
     val entryPointsByType: EntryPointsByType,
 
     val abi: List<AbiElement>? = null,
+) : ContractClassBase() {
+    @Serializable
+    data class EntryPointsByType(
+        @SerialName("CONSTRUCTOR")
+        val constructor: List<DeprecatedCairoEntryPoint>,
+
+        @SerialName("EXTERNAL")
+        val external: List<DeprecatedCairoEntryPoint>,
+
+        @SerialName("L1_HANDLER")
+        val l1Handler: List<DeprecatedCairoEntryPoint>,
+    )
+}
```

Comment:
In the newest starknet-spec (and devnet), `FunctionAbiEntry` also has an optional `stateVisibility` field, with only one allowed value, `view`.

Reply:
Oh, good catch, thanks. I have added this in newest commit + bumped devnet version